source-map.bs

```diff
@@ -274,6 +290,44 @@ fields represent mapped code that also has a mapped name.
 Note: Using file offsets was considered but rejected in favor of using line/column data to avoid becoming
 misaligned with the original due to platform-specific line endings. 
 
+### [=names=] for generated JavaScript code
+
+Source map generators should create a [=mapping=] entry with a [=name=] field for
+the following JavaScript tokens, if
+
+1. The original source language construct maps semantically to the generated
+    JavaScript code.
+
+1. The original source language construct has a name.
+
+Then the [=name=] of the [=mapping=] entry should be the name of the original
+source language construct.
+
+Example: A TypeScript arrow function compiled to a JavaScript function, or a
+local variable that was renamed.
+
+1. The opening parenthesis `(` of [=FunctionDeclaration=], [=FunctionExpression=],
+    [=AsyncFunctionDeclaration=] and [=AsyncFunctionExpression=] for (async) functions.
+
+1. The opening parenthesis `(` or the first letter of the *BindingExpression*
+    of [=ArrowParameters=] and [=AsyncArrowFunction=] for (async) arrow functions.
+
+1. The opening parenthesis `(` of [=MethodDefinition=] for class methods. For the
+    [=MethodDefinition=] where *PropName* is "constructor", the [=name=] should be
+    the original class name.
+
+1. The opening parenthesis `(` of [=GeneratorDeclaration=], [=GeneratorExpression=],
+    [=AsyncGeneratorDeclaration=] and [=AsyncGeneratorExpression=] for (async) generator functions.
+
+1. The first letter of [=Identifier=] for variables. This includes (but is not limited to)
+    [=LexicalDeclaration=], [=VariableStatement=] and arguments in [=Parameter List=].
+    Source map generators may choose to only emit a [=mapping=] entry with a [=name=] at a
+    variable's definition site and may choose to omit such a mapping for further use-sites
+    in the same [=Environment Record=].
```

Comment:
`Environment Record` is a _runtime_ concept, so it doesn't exist at parse/analysis time: maybe we should say "in the same scope"? Also, in
```js
function fn() {
  let x;
  x()
  function nested() { x() }
}
```
the two references to `x` are in two separate environment records. Do we need the "same environment record" restriction?

Reply:
Alternatively: We can also remove the distinction between definition site and use-site and recommend emitting name mappings generally for all `BindingIdentifier` in all expressions (so also all use-sites as well).

Reply:
Excellent point. I'm not that familiar with the official spec AST (yet).

So we could basically spec something like: "generators SHOULD emit a name mapping for `BindingIdentifier` and they MAY emit a name mapping for `IdentifierReference`?

Reply:
Yes, something like seems reasonable.

We could also recommend *not* to emit name mappings for `IdentifierReference`s, mostly because that would be bytes spent for no reason?